cpp/frontend_pyreadline.cc

```diff
@@ -235,4 +242,59 @@ Readline* MaybeGetReadline() {
 #endif
 }
 
+static void readline_cb(char* line) {
+  if (line == nullptr) {
+    gReadline->latest_line_ = nullptr;
+  } else {
+    gReadline->latest_line_ = line;
+  }
+  gReadline->ready_ = true;
+  rl_callback_handler_remove();
+}
+
+// See the following for some loose documentation on the approach here:
+// https://tiswww.case.edu/php/chet/readline/readline.html#Alternate-Interface-Example
+Str* readline(Str* prompt) {
+  fd_set fds;
+  FD_ZERO(&fds);
+  rl_callback_handler_install(prompt->data(), readline_cb);
+
+  gReadline->latest_line_ = nullptr;
+  gReadline->ready_ = false;
+  while (!gReadline->ready_) {
+    // Wait until stdin is ready or we are interrupted.
+    FD_SET(fileno(rl_instream), &fds);
+    int ec = select(FD_SETSIZE, &fds, NULL, NULL, NULL);
+    if (ec == -1) {
+      if (errno == EINTR && pyos::SigintCount() > 0) {
+        // User is trying to cancel. Abort and cleanup readline state.
+        rl_free_line_state();
+        rl_callback_sigcleanup();
+        rl_cleanup_after_signal();
+        rl_callback_handler_remove();
+        throw Alloc<KeyboardInterrupt>();
+      }
+
+      // To be consistent with CPython, retry on all other errors and signals.
+      continue;
+    }
+
+    // Remove this check if we start calling select() with a timeout above.
+    DCHECK(ec > 0);
+    if (FD_ISSET(fileno(rl_instream), &fds)) {
+      // Feed readline.
+      rl_callback_read_char();
+    }
+  }
+
+  if (gReadline->latest_line_ != nullptr) {
+    Str* s = StrFromC(gReadline->latest_line_);
```

Comment:
Yeah I like this new style where we do the `StrFromC()` allocation outside the callback

Even though it's a "synchronous callback" I like keeping all the GC stuff "on the edges" of the code, and having plain C in the  middle